articles/unique_network_flows.md

```diff
@@ -0,0 +1,171 @@
+
+## Problem
+
+All publishers and subscribers in communicating nodes have the same flow identifers (5-tuple or 3-tuple). This disables explicit network QoS differentiation for publishers and subscribers in communicating nodes. In other words, publishers and subscribers in communicating nodes can only be assigned the same network QoS.
+
+We believe the problem occurs by design. For performance reasons, RMW implementations are likely to associate IP address, port, and transport protocol to nodes and not to individual publishers/subscribers. This is true for all the tier-1 RMW implementations today (Foxy Fitzroy at the time of writing). None of the tier-1 RMW implementations set the IPv6 flow label to differentiate flows of publisher/subscribers.
+
+### Example
+
+We use the following example to highlight the problem.
+
+Consider a distributed robotics application with communicating nodes N1 and N2. N1 is active on device D1 and N2 on device D2. D1 and D2 are connected by an IP network, say a 5G network.
+
+N1 contains publishers P1 and P2. P1 publishes video data whereas P2 publishes battery status data.
+
+N2 contains subscribers S1 and S2. S1 receives video from P1 and performs real-time object detection. S2 receives battery data from P2 and performs non-real-time battery management.
+
+The link P1-S1 requires low-latency QoS from the 5G network, say a maximum delay of 5ms. P2-S2 requires default latency QoS i.e., 300ms. Then, by construction, since the flow identifiers of P1-S1 and P2-S2 links are the same, they cannot be assigned the required QoS by the network. Both P1-S1 and P2-S2 can either be assigned QoS with 5ms delay or with 300ms delay. The former case represents a waste of network resources, the latter case degrades performance.
+
+## Proposed Solution
+
+Our proposal to solve the problem is to make the flow identifiers of publishers and subscribers in communicating nodes unique using existing IP header fields.
+
+We believe there are at least three architectural options to implement unique network flows. These are shown in the figure below.
+
+![Architecture options for unique network flows](./unique-network-flows-arch-options.png)
+
+### Option 1: Unique flow ID from publisher/subscriber
+
+We construct a new QoS policy called *Network Flow* as a support structure to enable unique identification of flows.
+
+The Network Flow QoS policy is parameterized by just one parameter -- a 32-bit unsigned integer called `flow_id`. The default `flow_id` is zero.
+
+Each publisher or subscriber that requires a specific QoS from the network is required to set the `flow_id` parameter to value unique within the node. The `flow_id` can be computed internally in the node using simple UID generation schemes (atomic integer counters or random integer sampling without duplicates) or supplied by an external component.
```

Comment:
with this option, I think it would be better to provide API such as `generate_network_flow(void)` to user application to generate unique identification. if flow id depends on the implementation, it should ask implementation to allocate and return it to user application.

Reply:
Good suggestion! The delegation of unique flow ID generation to the RMW is covered by option-2 in the proposal (`network_flow(bool unique_flow = true)`).